services/src/main/java/org/keycloak/protocol/oidc/grants/device/endpoints/DeviceEndpoint.java

```diff
@@ -0,0 +1,488 @@
+
+        accessCode.setAction(AuthenticationSessionModel.Action.USER_CODE_VERIFICATION.name());
+        authenticationSession.setAuthNote(AuthenticationProcessor.CURRENT_AUTHENTICATION_EXECUTION, execution);
+
+        LoginFormsProvider provider = session.getProvider(LoginFormsProvider.class)
+            .setAuthenticationSession(authenticationSession).setExecution(execution)
+            .setClientSessionCode(accessCode.getOrGenerateCode());
+
+        if (errorMessage != null) {
+            provider = provider.setError(errorMessage);
+        }
+
+        return provider.createOAuth2DeviceVerifyUserCodePage();
+    }
+
+    public Response processVerification(AuthenticationSessionModel authenticationSession, String userCode) {
+        if (userCode == null) {
+            event.error(Errors.INVALID_OAUTH2_USER_CODE);
+            return createVerificationPage(Messages.OAUTH2_DEVICE_INVALID_USER_CODE, authenticationSession);
+        }
+
+        // Format inputted user code
+        OAuth2DeviceUserCodeProvider userCodeProvider = session.getProvider(OAuth2DeviceUserCodeProvider.class);
+        String formattedUserCode = userCodeProvider.format(userCode);
+
+        // Find the token from store
+        OAuth2DeviceTokenStoreProvider store = session.getProvider(OAuth2DeviceTokenStoreProvider.class);
+        OAuth2DeviceCodeModel deviceCodeModel = store.getByUserCode(realm, formattedUserCode);
+        if (deviceCodeModel == null) {
+            event.error(Errors.INVALID_OAUTH2_USER_CODE);
+            return createVerificationPage(Messages.OAUTH2_DEVICE_INVALID_USER_CODE, authenticationSession);
+        }
+
+        int expiresIn = deviceCodeModel.getExpiration() - Time.currentTime();
+        if (expiresIn < 0) {
+            event.error(Errors.EXPIRED_OAUTH2_USER_CODE);
+            return createVerificationPage(Messages.OAUTH2_DEVICE_INVALID_USER_CODE, authenticationSession);
+        }
+
+        // Verification OK
+        authenticationSession.setClientNote(DeviceGrantType.OAUTH2_DEVICE_VERIFIED_USER_CODE, formattedUserCode);
+
+        // Event logging for the verification
+        event.client(deviceCodeModel.getClientId()).detail(Details.SCOPE, deviceCodeModel.getScope()).success();
+
+        OIDCLoginProtocol protocol = new OIDCLoginProtocol(session, realm, session.getContext().getUri(), headers, event);
+        return handleBrowserAuthenticationRequest(authenticationSession, protocol, false, true);
+    }
+
+    @Override
+    public Object getResource() {
+        return this;
+    }
+
+    @Override
+    public void close() {
+
+    }
+
+    private ClientModel authenticateClient() {
+        // https://tools.ietf.org/html/draft-ietf-oauth-device-flow-15#section-3.1
+        // The spec says "The client authentication requirements of Section 3.2.1 of [RFC6749]
+        // apply to requests on this endpoint".
+        AuthorizeClientUtil.ClientAuthResult clientAuth = AuthorizeClientUtil.authorizeClient(session, event);
```

Comment:
I think it will be needed to rebase this PR. In latest master, signature of AuthorizeClientUtil.authorizeClient slightly changed.

Reply:
I fixed.